lib/api/bucketPutPolicy.js

```diff
@@ -0,0 +1,71 @@
+const async = require('async');
+const { BucketPolicy } = require('arsenal').models;
+
+const collectCorsHeaders = require('../utilities/collectCorsHeaders');
+const metadata = require('../metadata/wrapper');
+const { metadataValidateBucket } = require('../metadata/metadataUtils');
+const { pushMetric } = require('../utapi/utilities');
+const { errors } = require('arsenal');
+
+/**
+ * bucketPutPolicy - create or update a bucket policy
+ * @param {AuthInfo} authInfo - Instance of AuthInfo class with requester's info
+ * @param {object} request - http request object
+ * @param {object} log - Werelogs logger
+ * @param {function} callback - callback to server
+ * @return {undefined}
+ */
+function bucketPutPolicy(authInfo, request, log, callback) {
+    log.debug('processing request', { method: 'bucketPutPolicy' });
+
+    const { bucketName } = request;
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        requestType: 'bucketOwnerAction',
+    };
+
+    if (!process.env.BUCKET_POLICY) {
+        return callback(errors.NotImplemented);
+    }
+
+    return async.waterfall([
+        next => {
+            const policyJSON = JSON.parse(request.post);
+            const bucketPolicy = new BucketPolicy(policyJSON);
+            // if there was an error getting bucket policy,
+            // returned configObj will contain 'error' key
+            process.nextTick(() => {
+                const policyObj = bucketPolicy.getBucketPolicy();
+                return next(policyObj.error || null, policyObj);
```

Comment:
If the policy object has an error, I'm wondering whether you'll need to find a way to pass a `BucketInfo` instance for the final callback in the waterfall. I'm thinking that `collectCorsHeaders` will error when trying to call `getCors` [here](https://github.com/scality/cloudserver/blob/development/8.1/lib/utilities/collectCorsHeaders.js#L24).

Reply:
That's a good catch, but the `collectCorsHeaders` function just returns an empty object if no bucket is passed to it :)

Reply:
Isn't `policyObj` passed as `bucket` to that function?

Reply:
No, if there's no error, `policyObj` is just passed to the next function in the waterfall. The var name changes to `bucketPolicy` in the next function, but then that's where the `bucket` var is created

Reply:
I mean if there is an error. As in, `policyObj.error` is truthy.